packages/apollo/src/utils/lists.ts

```diff
@@ -67,26 +106,58 @@ export class ListHelper {
             this.skipCacheOnRead = options.cacheOptions.skipOnRead
         }
 
-        bindAll(this, '_readPage', '_networkOnlyRead', 'getReadFunction', 'mergeLists')
+        bindAll(this, '_readPage', '_readList', '_networkOnlyRead', 'getReadFunction', 'mergeLists')
     }
 
     /**
-     * Read function, which implements paginated reading from cache
+     * Read function that implements paginated reading from cache.
+     * Returns a slice of the cached list based on pagination arguments.
+     * Includes validation of data integrity by checking for broken references
+     * (references to objects that don't exist in the cache).
      */
     private _readPage<TData, TOptions extends FieldFunctionOptions>(
         existing: ReadonlyArray<TData> | undefined,
         options: TOptions
     ): Array<TData> | undefined {
-        const skip: number = get(options, ['args', this.skipArgName], 0)
-        const first: number = get(options, ['args', this.firstArgName], 0)
+        const skip: number = options?.args?.[this.skipArgName] || 0
+        const first: number = options?.args?.[this.firstArgName] || 0
 
         // NOTE: partial data (skip only) is not enough for case
         // where you fetch (page + 1) item to show "Next page" pagination event
         if (!existing || existing.length < skip + first) {
             return undefined
         }
 
-        return existing.slice(skip, skip + first)
+        const filteredPage = existing.slice(skip, skip + first).filter(Boolean)
```

Comment:
We should filter only `null` / `undefined`, so if your return type is [Number], you can filter out zeros...

Reply:
It's better be more specific and use `nonNull` here:
```
import { nonNull } from '@open-condo/miniapp-utils/helpers/collections'
```